reconcile/rhidp/ocm_oidc_idp/integration.py

```diff
@@ -0,0 +1,58 @@
+import logging
+import sys
+from typing import (
+    Any,
+    Callable,
+    Optional,
+)
+
+from reconcile.gql_definitions.rhidp.clusters import ClusterV1
+from reconcile.rhidp.common import get_clusters
+from reconcile.rhidp.ocm_oidc_idp.base import run
+from reconcile.status import ExitCodes
+from reconcile.utils import gql
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+
+QONTRACT_INTEGRATION = "ocm-oidc-idp"
+
+
+class OCMOidcIdpIntegrationParams(PydanticRunParams):
+    vault_input_path: str
+
+
+class OCMOidcIdpIntegration(
+    QontractReconcileIntegration[OCMOidcIdpIntegrationParams],
+):
+    """A flavour of the OCM OIDC IDP integration, that receives the list of
+    clusters from app-interface.
+    """
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def run(self, dry_run: bool) -> None:
+        gqlapi = gql.get_api()
+        clusters = self.get_clusters(gqlapi.query)
+        # data query
+        if not clusters:
+            logging.debug("No clusters with oidc-idp definitions found.")
+            sys.exit(ExitCodes.SUCCESS)
+
+        run(
+            integration_name=self.name,
+            clusters=clusters,
+            secret_reader=self.secret_reader,
+            vault_input_path=self.params.vault_input_path,
+            dry_run=dry_run,
+        )
+
+    def get_clusters(self, query_func: Callable) -> list[ClusterV1]:
```

Comment:
If here pass `gqlapi` instead of `query_func`, it will be easier to setup mock in test, what do you think?

Reply:
TBH, I don't like mocking. Passing a `query_func` makes testing easier without mocking. This is a common pattern, used in other test cases too.

Reply:
Mocking is needed anyway, but in different layers, it can be a manual one like [this](https://github.com/app-sre/qontract-reconcile/pull/3598/files#diff-551e1d77b15ca2654a18f6b7b8e257ec10fcdccc963dd9f6a21d8678312a74b9R29), but it will miss some handy assertions like `assert_called_once_with`. I also don't like mocking, I like classical TDD and sociable unit tests, [ref](https://martinfowler.com/articles/practical-test-pyramid.html#SociableAndSolitary).

Just clarify what I mean `gqlapi` is easier than `query_func`, if it's

```python
    def get_clusters(self, gqlapi: gql.GqlApi) -> list[ClusterV1]:
        return get_clusters(self.name, gqlapi.query)
```

Then unit test can be done like this

```python
from unittest.mock import create_autospec

mock_gqlapi = create_autospec(spec=gqlapi.GqlApi)
mock_get_clusters = mocker.patch('reconcile.rhidp.ocm_oidc_idp.integration.py.get_clusters')
intg = OCMOidcIdpIntegration()

intg.get_clusters(mock_gqlapi)

mock_get_clusters.assert_called_once_with("ocm-oidc-idp", mock_gqlapi.query)
```

Another part makes this method hard to test is the free var `get_clusters` from `from reconcile.rhidp.common import get_clusters`, that requires a `mocker.patch` to test this method in solitary.

Your tests call through `get_clusters`, only the last layer is mocked, so there is no problem with it.

I tried to use dependency injection way of passing `query_func`, but mypy is not happy about it, that's why in the end I switched to passing `GqlApi` object, https://github.com/app-sre/qontract-reconcile/pull/3467#discussion_r1173829008